text/0000-new-struct-syntax.md

```diff
@@ -0,0 +1,188 @@
+```
+
+Is `{a: b}` a struct literal with value `b` mapped to `a` or is it a code block with expression `a: b` (`a` with ascribed type `b`)?
+
+## What would be the ideal syntax?
+
+There are few important characteristics that a struct literal syntax should have:
+
+1.	It should not interfere with any existing operator that is currently (or might be in the future) used for different thing.
+
+2.	It should conform to the initialization-follows-declaration rule so that the same syntax is used when declaring struct type and specifying a struct value.
+
+3.	It should be easily distinguished from a code block so that single-value struct literal and single-statement code block are distinct.
+
+4.	It should fit in a pattern naturally, and not clash with any operators used there (important for future named arguments feature).
+
+# Detailed design
+
+The proposal is to simply replace the `:` in structs with `=>`.
+`=>` would then become the key-value mapping operator.
+It is currently used in match blocks where it serves a similar purpose.
+
+## Usage in expressions
+
+Take the current definition of struct literal expressed roughly using PEG syntax:
+
+```peg
+expression ::= identifier "{" ( identifier ":" expression "," )* "}"
+```
+
+The new syntax would then become:
+
+```peg
+expression ::= identifier "{" ( identifier "=>" expression "," )* "}"
+```
+
+Which combined with type ascription (which extends `expression` with `expression ":" type`)
+could allow us to instantiate a struct value with this syntax:
+
+```rust
+let p = Point { x => 1: f32, y => 2 }
+```
+
+## Usage in patterns
+
+The current syntax of a struct in the context of a pattern could be defined as follows:
+
+```peg
+pattern ::= identifier "{" ( identifier ":" pattern "," )* "}"
+```
+
+The new syntax would then become:
+
+```peg
+pattern ::= identifier "{" ( identifier "=>" pattern "," )* "}"
+```
+
+Combined with type ascription (which extends `pattern` with `pattern ":" type`)
+would allow patterns like:
+
+```rust
+let Point { x => a: f32, y => b } = get_origin();
+// variables `a` and `b` now hold the coordinates
+```
+
+If `get_point` was generic, the `f32` type ascription on `x` field forces it to return `Point<f32>`.
+Because pattern or subpattern can be specified as a variable name that gets the matched value assigned to,
+destructuring syntax with type ascription becomes `{ field => variable: Type }`
+
+## Declaration of struct type
+
+There's nothing wrong with the current declaratiom syntax,
+but leaving it unchanged while changing the other syntax would make structs
+the only data type which doesn't conform to the initialization-follows-declaration rule.
+To preserve the symmetry, the proposal is to change `:` to `=>` in struct declaration aswell.
+
+```rust
+struct Color<T>(T, T, T); // declaration
+set_color(Color(0: u8, 0, 0)); // initialization
+let Color(red: u8, grn, blu) = get_color(); // pattern matching
+
+struct Color<T>{r=>T, g=>T, b=>T}; // declaration
```

Comment:
Ya, I'm not sure why `=>` is used as "has type" here. Wasn't the idea to have `:` mean "has type" everywhere?